text/0000-feature-metadata.md

```diff
@@ -0,0 +1,309 @@
+there are some limitations that don't have elegant solutions:
+
+-   Documentation is difficult, often requiring library authors to manually manage
+    a table of descriptions
+-   There is no way to deprecate old features, as a way to help crates maintain
+    semvar compliance
+-   Features cannot be hidden from use in any way
+
+This RFC proposes a plan that add that information to `Cargo.toml`, solving
+these problems.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Usage is simple: features will be able to be specified as a table, instead of
+just a dependency array. This sample section of `Cargo.toml` shows new
+possibilities:
+
+```toml
+[features]
+# Current configuration will continue to work
+foo = []
+# New configurations
+bar = { requires = ["foo"], doc = "simple docstring here"}
+baz = { requires = ["foo"], public = false}
+qux = { requires = [], deprecated = true, unstable = true }
+quux = { requires = [], deprecated = { since = "1.2.3", note = "don't use this!" } }
+
+# Features can also be full tables if descriptions are longer
+[features.corge]
+requires = ["bar", "baz"]
+doc = """
+# corge
+
+This could be a longer description of this feature
+"""
+```
+
+The following keys would be allowed in a feature table:
+
+-   `requires`: This is synonymous with the existing array describing required
+    features. For example, `foo = ["dep:serde", "otherfeat"]` will be identical to
+    `foo = { requires = ["dep:serde", "otherfeat"] }`
+-   `doc`: A markdown docstring describing the feature. Like with `#[doc(...)]`,
+    the first line will be treated as a summary.
+-   `deprecated`: This can be either a simple boolean, a string, or an object
+    with `since` and/or `note` keys. Cargo will warn downstream crates using this
+-   `public`: A boolean flag defaulting to `true` that indicates whether or not
+    downstream crates should be allowed to use this feature
+-   `unstable`: A boolean flag indicating that the feature is only usable with
+    nightly Rust, or is otherwise not API-stable
+
+If a downstream crate attempts to use the features `baz` and `qux`, they will
+see messages like the following:
+
+```
+warning: feature `quux` on crate `mycrate` is deprecated since version
+  1.2.3: "don't use this!"
+
+error: feature `baz` on crate `mycrate` is private and cannot be used by
+  downstream crates
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Validation and parsing of the new schema, described above, shoud be relatively
+straightforward. Each of the added keys is discussed individually in the
+following sections:
+
+## `requires`
+
+`requires` will take the place of the feature dependency array that currently
+exists. Semantics will remain unchanged.
+
+This is a required key.
+
+## `doc`
+
+`doc` is the most straightforward: it accepts markdown-flavored text, and should
```

Comment:
I have it as-is because of the original intent to get this information to rustdoc somehow, where formatting would be nice. Would matching `package.description` be for the point of consistency, or to avoid needing a markdown parser?

I would also be OK not doing markdown in the initial proposal and maybe adding that later if that would be better. I'll defer to you for what would be best here.

(fwiw I restarted work on my markdown->term tool for rust explain https://github.com/rust-lang/rust/pull/112697. The parser isn't compliant so Cargo couldn't use it, but it would be easy to use `pulldown-cmark` and keep the printing so things would be consistent.)

Reply:
> Would matching package.description be for the point of consistency, or to avoid needing a markdown parser?

The question of what we should be consistent with either needs to be resolved and documented in alternatives or listed in Unresolved Questions

Reply:
I did move this to an unresolved question, with the statement that we can make this decision at implementation time once we have a better idea of how difficult markdown support is. I'll mark this as resolved, reopen if needed.